server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/projection/cassandra/EnqueuedMailsDao.java

```diff
@@ -75,9 +76,111 @@
+            Date lastUpdated = row.getTimestamp(LAST_UPDATED);
+            Map<String, ByteBuffer> rawAttributes = row.getMap(ATTRIBUTES, String.class, ByteBuffer.class);
+            PerRecipientHeaders perRecipientHeaders = fromHeaderMap(row.getMap(PER_RECIPIENT_SPECIFIC_HEADERS, String.class, UDTValue.class));
+
+            MailImpl mail = MailImpl.builder()
+                .name(name)
+                .sender(sender)
+                .recipients(recipients)
+                .lastUpdated(lastUpdated)
+                .errorMessage(errorMessage)
+                .remoteHost(remoteHost)
+                .remoteAddr(remoteAddr)
+                .state(state)
+                .addAllHeadersForRecipients(perRecipientHeaders)
+                .attributes(toAttributes(rawAttributes))
+                .build();
+
+            ManageableMailQueue.MailQueueItemView mailQueueItemView = new ManageableMailQueue.MailQueueItemView(mail);
+            return new CassandraMailQueueProjectionItem(queueName, MailKey.of(name), mailQueueItemView);
+        }
+
+        private static Map<String, Serializable> toAttributes(Map<String, ByteBuffer> rowAttributes) {
+            return rowAttributes.entrySet()
+                .stream()
+                .map(entry -> Pair.of(entry.getKey(), fromByteBuffer(entry.getValue())))
+                .collect(Guavate.toImmutableMap(Pair::getLeft, Pair::getRight));
+        }
+
+        private static Serializable fromByteBuffer(ByteBuffer byteBuffer) {
+            try {
+                byte[] data = new byte[byteBuffer.remaining()];
+                byteBuffer.get(data);
+                ObjectInputStream objectInputStream = new ObjectInputStream(new ByteArrayInputStream(data));
+                return (Serializable) objectInputStream.readObject();
+            } catch (IOException | ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private static PerRecipientHeaders fromHeaderMap(Map<String, UDTValue> rawMap) {
+            PerRecipientHeaders result = new PerRecipientHeaders();
+
+            rawMap.forEach((key, value) -> result.addHeaderForRecipient(PerRecipientHeaders.Header.builder()
+                    .name(value.getString(HEADER_NAME))
+                    .value(value.getString(HEADER_VALUE))
+                    .build(),
+                toMailAddress(key)));
+            return result;
+        }
+
+        private static MailAddress toMailAddress(String rawValue) {
+            try {
+                return new MailAddress(rawValue);
+            } catch (AddressException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    static class BucketAndSlice {
+        static BucketAndSlice of(Integer bucketId, Long sliceTime) {
+            return new BucketAndSlice(bucketId, sliceTime);
+        }
+
+        private final Integer bucketId;
+        private final Date sliceTime;
```

Comment:
you should not rely on Date in new code, use java.time instead

Reply:
I find it's convenient to use `Date` as an input of cassandra binding apis, then I take it as a `Date`.

Reply:
We should not use Date anymore because (main reason) it doesn't carry its timezone so it's not an Instant. There's a lot of other reasons too but a comment is not really the right place for writing a book. This code being purely technical, we generate the date and should not take care of the timezone so we don't have to deal with DateTime but I think that Instant is representing our concept much better.

Reply:
As I remember well, `Date` also represents the UTC time zone milliseconds. Anw, I will try to remove the `Date`